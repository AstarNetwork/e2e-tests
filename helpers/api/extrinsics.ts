import { AccountInfo } from '@polkadot/types/interfaces'
import { ApiPromise } from '@polkadot/api'
import { SubmittableExtrinsic } from '@polkadot/api-base/types'

export const balance = async (api: ApiPromise, address: string) => {
  const account = await api.query.system.account<AccountInfo>(address)
  return account.data.toJSON()
}

export const xTokensForRelayChain = (api: ApiPromise, amount: string, address: Uint8Array) => {
  return api.tx.xTokens.transfer(
    {
      Token: 'KSM',
    },
    amount,
    {
      V1: {
        parents: 1,
        interior: {
          X1: {
            AccountId32: {
              network: 'Any',
              id: address,
            },
          },
        },
      },
    },
    'Unlimited',
  )
}

export const xTokensForParaChain = (
  api: ApiPromise,
  token: object,
  parachainId: string,
  amount: string,
  address: Uint8Array,
) => {
  return api.tx.xTokens.transfer(
    token,
    amount,
    {
      V1: {
        parents: 1,
        interior: {
          X2: [
            {
              Parachain: parachainId,
            },
            {
              AccountId32: {
                network: 'Any',
                id: address,
              },
            },
          ],
        },
      },
    },
    'Unlimited',
  )
}

export const xTokens = (
  api: ApiPromise,
  isRelayChain: boolean,
  parachainId: string,
  token: object,
  amount: string | bigint,
  address: Uint8Array,
) => {
  const multiLocation = isRelayChain
    ? { X1: { AccountId32: { network: 'Any', id: address } } }
    : {
        X2: [
          {
            Parachain: parachainId,
          },
          {
            AccountId32: {
              network: 'Any',
              id: address,
            },
          },
        ],
      }

  return api.tx.xTokens.transfer(
    token,
    amount,
    {
      V1: {
        parents: 1,
        interior: multiLocation,
      },
    },
    'Unlimited',
  )
}

export const xTokensV3 = (
  api: ApiPromise,
  isRelayChain: boolean,
  parachainId: string,
  token: object,
  amount: string | bigint,
  address: Uint8Array,
) => {
  const multiLocation = isRelayChain
    ? {
        X1: {
          AccountId32: {
            id: address,
          },
        },
      }
    : {
        X2: [
          {
            Parachain: parachainId,
          },
          {
            AccountId32: {
              id: address,
            },
          },
        ],
      }

  return api.tx.xTokens.transfer(
    token,
    amount,
    {
      V3: {
        parents: 1,
        interior: multiLocation,
      },
    } as any,
    'Unlimited',
  )
}

export const relayChainV3limitedReserveTransferAssets = (
  api: ApiPromise,
  parachainId: string,
  amount: string,
  address: Uint8Array,
) => {
  return api.tx.xcmPallet.limitedReserveTransferAssets(
    {
      V3: {
        parents: 0,
        interior: {
          X1: { Parachain: parachainId },
        },
      },
    },
    {
      V3: {
        parents: 0,
        interior: {
          X1: {
            AccountId32: {
              id: address,
            },
          },
        },
      },
    },
    {
      V3: [
        {
          id: { Concrete: { parents: 0, interior: 'Here' } },
          fun: { Fungible: amount },
        },
      ],
    },
    0,
    'Unlimited',
  )
}

export const xTokensTransferMulticurrencies = (
  api: ApiPromise,
  foreignAssetId: string,
  amount: string,
  parachainId: string,
  address: Uint8Array,
) => {
  return api.tx.xTokens.transferMulticurrencies(
    [
      [
        {
          ForeignAsset: foreignAssetId,
        },
        amount,
      ],
      [
        {
          Token: 'KSM',
        },
        '16000000000',
      ],
    ],
    '1',
    {
      V1: {
        parents: 1,
        interior: {
          X2: [
            {
              Parachain: parachainId,
            },
            {
              AccountId32: {
                network: 'Any',
                id: address,
              },
            },
          ],
        },
      },
    },
    'Unlimited',
  )
}

export const xTokensTransferTransferMultiasset = (
  api: ApiPromise,
  interior: any[],
  amount: string,
  parachainId: string,
  address: Uint8Array,
) => {
  return api.tx.xTokens.transferMultiasset(
    {
      V1: {
        fun: {
          Fungible: amount,
        },
        id: {
          Concrete: {
            parents: 1,
            interior: {
              X3: interior,
            },
          },
        },
      },
    },
    {
      V1: {
        parents: 1,
        interior: {
          X2: [
            {
              Parachain: parachainId,
            },
            {
              AccountId32: {
                network: 'Any',
                id: address,
              },
            },
          ],
        },
      },
    },
    'Unlimited',
  )
}

export const xTokensTransferTransferMultiassetV3 = (
  api: ApiPromise,
  interior: any[],
  amount: string,
  parachainId: string,
  address: Uint8Array,
) => {
  return api.tx.xTokens.transferMultiasset(
    {
      V3: {
        fun: {
          Fungible: amount,
        },
        id: {
          Concrete: {
            parents: 1,
            interior: {
              X3: interior,
            },
          },
        },
      },
    },
    {
      V3: {
        parents: 1,
        interior: {
          X2: [
            {
              Parachain: parachainId,
            },
            {
              AccountId32: {
                id: address,
              },
            },
          ],
        },
      },
    } as any,
    'Unlimited',
  )
}

export const xTokensTransferMulticurrenciesV3 = (
  api: ApiPromise,
  foreignAssetId: string,
  amount: string,
  parachainId: string,
  address: Uint8Array,
) => {
  return api.tx.xTokens.transferMulticurrencies(
    [
      [
        {
          ForeignAsset: foreignAssetId,
        },
        amount,
      ],
      [
        {
          Token: 'KSM',
        },
        '16000000000',
      ],
    ],
    '1',
    {
      V3: {
        parents: 1,
        interior: {
          X2: [
            {
              Parachain: parachainId,
            },
            {
              AccountId32: {
                id: address,
              },
            },
          ],
        },
      },
    } as any,
    'Unlimited',
  )
}

export const sudo = (api: ApiPromise, call: SubmittableExtrinsic<'promise'>) => {
  return api.tx.sudo.sudo(call)
}
