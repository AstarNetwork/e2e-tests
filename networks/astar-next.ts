import { Config } from './types'

import astarConfig, { Vars } from './astar'

export default {
  ...astarConfig,
  polkadot: {
    ...astarConfig.polkadot,
    name: 'astarNext' as const,
  },
  kusama: {
    ...astarConfig.kusama,
    name: 'shidenNext' as const,
  },
  config: (opt) => ({
    ...astarConfig.config(opt),
    options: {
      wasmOverride: {
        polkadot: './wasm/astar_runtime.wasm',
        kusama: './wasm/shiden_runtime.wasm',
      }[opt.network],
    },
  }),
} satisfies Config<Vars>
