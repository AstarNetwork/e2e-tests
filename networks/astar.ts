import { Config } from './types'

export type Vars = {
  relayToken: string
}

export default {
  polkadot: {
    name: 'astar' as const,
    endpoint: 'wss://astar.api.onfinality.io/public-ws',
    relayToken: '340282366920938463463374607431768211455',
  },
  kusama: {
    name: 'shiden' as const,
    endpoint: 'wss://shiden.api.onfinality.io/public-ws',
    relayToken: '340282366920938463463374607431768211455',
  },
  config: ({ alice, relayToken }) => ({
    storages: {
      System: {
        Account: [[[alice.address], { data: { free: 1000n * 10n **18n } }]],
      },
      Assets: {
        account: [[[relayToken, alice.address], { balance: 10n ** 18n }]],
      },
      Sudo: {
        key: alice.address,
      },
      PolkadotXcm: {
        // avoid sending xcm version change notifications to makes things faster
        $removePrefix: ['versionNotifyTargets', 'versionNotifiers', 'supportedVersion'],
      },
    },
  }),
} satisfies Config<Vars>

export const astar = {
  paraId: 2006,
  paraAccount: '13YMK2eZzuFY1WZGagpYtTgbWBWGdoUD2CtrPj1mQPjY8Ldc',
  dot: 340282366920938463463374607431768211455n,
  astr: { Concrete: { parents: 0, interior: 'Here' } },
  aca: 18446744073709551616n,
  usdt: 4294969280n,
  dot_loc : { Concrete: { parents: 1, interior: 'Here' } },
  usdt_loc: { Concrete: { parents: 1, interior: { X3: [ { Parachain: 1000 }, { PalletInstance: 50 }, { GeneralIndex: 1984 }] } } },
} as const

export const shiden = {
  paraId: 2007,
  paraAccount: 'F7fq1jNy74AqkJ1DP4KqSrWtnTGtXfNVoDwFhTvvPxUvJaq',
  ksm: 340282366920938463463374607431768211455n,
  sdn: { Concrete: { parents: 0, interior: 'Here' } },
  kar: 18446744073709551618n,
  usdt: 4294969280n,
  ausd: 18446744073709551616n,
} as const
