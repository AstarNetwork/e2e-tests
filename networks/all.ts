import { Config } from './types'

import astar from './astar'
import astarNext from './astar-next'
import polkadot from './polkadot'
import statemint from './statemint'

const all = {
  polkadot,
  statemint,
  astar,
  astarNext,
} satisfies Record<string, Config>

export default all
