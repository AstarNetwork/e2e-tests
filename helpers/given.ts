import { test } from 'vitest'
import _ from 'lodash'

import { Network, NetworkNames, createContext, createNetworks } from '../networks'

class NetworkManager {
  keyring = createContext()

  async buildNetworks(names: NetworkNames[]) {
    const networkOptions = _.zipObject(_.zip(names) as any) as any as Record<NetworkNames, undefined>
    return createNetworks(networkOptions, this.keyring)
  }
}

export const given = <T extends NetworkNames>(...names: NetworkNames[]) =>
  test.extend<{
    networks: Record<T, Network>
    keyring: ReturnType<typeof createContext>
  }>({
    networks: async ({ task }, use) => {
      if (task.retry) {
        throw new Error('given cannot run with retry')
      }
      if (task.concurrent) {
        throw new Error('given cannot run in concurrent mode')
      }
      const manager = new NetworkManager()
      _.assign(task.suite, { __networkManager__: manager })
      await use(await manager.buildNetworks(names))
    },
    keyring: async ({ task }, use) => {
      // network manager is assigned above
      const manager: NetworkManager = _.get(task.suite, '__networkManager__')!
      await use(manager.keyring)
    },
  })
