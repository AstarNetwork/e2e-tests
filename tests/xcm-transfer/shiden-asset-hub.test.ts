import { checkHrmp, checkSystemEvents, given } from '../../helpers'
import { describe, expect } from 'vitest'
import { query, tx } from '../../helpers/api'
import { shiden as shidenUtil } from '../../networks/astar'
import { statemine as statemineUtil } from '../../networks/statemint'

describe('Shiden & AssetHub', () => {
  given('shiden', 'statemine')(
    '001: AssetHub transfer USDT to Shiden',
    async ({ networks: { shiden, statemine }, keyring: { alice, bob } }) => {
      await statemine.dev.setStorage({
        Assets: {
          account: [[[statemineUtil.usdtIndex, alice.address], { balance: 100e6 }]],
        },
      })

      await tx.xcmPallet
        .limitedReserveTransferAssetsV3(statemineUtil.usdt, 10e6, tx.xcmPallet.parachainV3(
            1,
            shidenUtil.paraId,
          ))(statemine, bob.addressRaw)
        .signAndSend(alice)

      await statemine.chain.newBlock()

      await checkHrmp(statemine)
        .redact({ redactKeys: /setTopic/ })
        .toMatchSnapshot('001: statemine ump messages')

      await shiden.chain.newBlock()

      const bobBalance = await query.assets(shidenUtil.usdt)(shiden, bob.address)
      expect(bobBalance.unwrap().balance.toNumber()).closeTo(
        10_000_000,
        10_000, // some fee
        'Expected amount was not received',
      )

      await checkSystemEvents(shiden, 'xcmpQueue', 'messageQueue').toMatchSnapshot('002: shiden event')
    },
  )

  given('shiden', 'statemine')(
    '002: Shiden transfer USDT to AssetHub',
    async ({ networks: { shiden, statemine }, keyring: { alice, bob } }) => {
      await shiden.dev.setStorage({
        Assets: {
          account: [[[shidenUtil.usdt, alice.address], { balance: 100e6 }]],
        },
      })

      await tx.xtokens
        .transfer(shidenUtil.usdt, 10e6, tx.xtokens.parachainV3(statemineUtil.paraId))(shiden, bob.addressRaw)
        .signAndSend(alice)

      await shiden.chain.newBlock()

      await checkHrmp(shiden)
        .redact({ redactKeys: /setTopic/ })
        .toMatchSnapshot('001: shiden ump messages')

      await statemine.chain.newBlock()

      const bobBalance = await query.assets(statemineUtil.usdtIndex)(statemine, bob.address)
      expect(bobBalance.unwrap().balance.toNumber()).closeTo(
        10_000_000,
        20_000, // some fee
        'Expected amount was not received',
      )

      await checkSystemEvents(statemine, 'xcmpQueue', 'messageQueue').toMatchSnapshot('002: statemine event')
    },
  )

  given('shiden', 'statemine')(
    '003: Transfer KSM from AssetHub (statemine) to Shiden',
    async ({ networks: { shiden, statemine }, keyring: { alice, bob } }) => {
      await tx.xcmPallet
        .limitedReserveTransferAssetsV3(statemineUtil.ksm, 1e12, tx.xcmPallet.parachainV3(
          1,
          shidenUtil.paraId,
        ))(statemine, bob.addressRaw)
        .signAndSend(alice)

      await statemine.chain.newBlock()

       await checkHrmp(statemine)
        .redact({ redactKeys: /setTopic/ })
        .toMatchSnapshot('003: statemine ump messages')

      await shiden.chain.newBlock()

      const bobBalance = await query.assets(shidenUtil.ksm)(shiden, bob.address)
      expect(bobBalance.unwrap().balance.toNumber()).closeTo(
        1_000_000_000_000,
        1_000_000, // some fee
        'Expected amount was not received',
      )

      await checkSystemEvents(shiden, 'xcmpQueue', 'messageQueue').toMatchSnapshot('003: shiden event')
    },
  )
})
