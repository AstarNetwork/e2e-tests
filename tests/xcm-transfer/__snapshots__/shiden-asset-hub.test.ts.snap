// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Shiden & AssetHub > 001: AssetHub transfer USDT to Shiden > 001: statemine ump messages 1`] = `
[
  {
    "data": [
      "ConcatenatedVersionedXcm",
      {
        "v4": [
          {
            "reserveAssetDeposited": [
              {
                "fun": {
                  "fungible": 10000000,
                },
                "id": {
                  "interior": {
                    "x3": [
                      {
                        "parachain": 1000,
                      },
                      {
                        "palletInstance": 50,
                      },
                      {
                        "generalIndex": 1984,
                      },
                    ],
                  },
                  "parents": 1,
                },
              },
            ],
          },
          {
            "clearOrigin": null,
          },
          {
            "buyExecution": {
              "fees": {
                "fun": {
                  "fungible": 10000000,
                },
                "id": {
                  "interior": {
                    "x3": [
                      {
                        "parachain": 1000,
                      },
                      {
                        "palletInstance": 50,
                      },
                      {
                        "generalIndex": 1984,
                      },
                    ],
                  },
                  "parents": 1,
                },
              },
              "weightLimit": {
                "unlimited": null,
              },
            },
          },
          {
            "depositAsset": {
              "assets": {
                "wild": {
                  "allCounted": 1,
                },
              },
              "beneficiary": {
                "interior": {
                  "x1": [
                    {
                      "accountId32": {
                        "id": "0xd17c2d7823ebf260fd138f2d7e27d114c0145d968b5ff5006125f2414fadae69",
                        "network": null,
                      },
                    },
                  ],
                },
                "parents": 0,
              },
            },
          },
          {
            "setTopic": "(redacted)",
          },
        ],
      },
    ],
    "recipient": 2007,
  },
]
`;

exports[`Shiden & AssetHub > 001: AssetHub transfer USDT to Shiden > 002: shiden event 1`] = `
[
  {
    "data": {
      "id": "(hash)",
      "origin": {
        "Sibling": 1000,
      },
      "success": true,
      "weightUsed": {
        "proofSize": "(rounded 9300)",
        "refTime": "(rounded 880000000)",
      },
    },
    "method": "Processed",
    "section": "messageQueue",
  },
]
`;

exports[`Shiden & AssetHub > 002: Shiden transfer USDT to AssetHub > 001: shiden ump messages 1`] = `
[
  {
    "data": [
      "ConcatenatedVersionedXcm",
      {
        "v2": [
          {
            "withdrawAsset": [
              {
                "fun": {
                  "fungible": 10000000,
                },
                "id": {
                  "concrete": {
                    "interior": {
                      "x2": [
                        {
                          "palletInstance": 50,
                        },
                        {
                          "generalIndex": 1984,
                        },
                      ],
                    },
                    "parents": 0,
                  },
                },
              },
            ],
          },
          {
            "clearOrigin": null,
          },
          {
            "buyExecution": {
              "fees": {
                "fun": {
                  "fungible": 10000000,
                },
                "id": {
                  "concrete": {
                    "interior": {
                      "x2": [
                        {
                          "palletInstance": 50,
                        },
                        {
                          "generalIndex": 1984,
                        },
                      ],
                    },
                    "parents": 0,
                  },
                },
              },
              "weightLimit": {
                "unlimited": null,
              },
            },
          },
          {
            "depositAsset": {
              "assets": {
                "wild": {
                  "all": null,
                },
              },
              "beneficiary": {
                "interior": {
                  "x1": {
                    "accountId32": {
                      "id": "0xd17c2d7823ebf260fd138f2d7e27d114c0145d968b5ff5006125f2414fadae69",
                      "network": {
                        "any": null,
                      },
                    },
                  },
                },
                "parents": 0,
              },
              "maxAssets": 1,
            },
          },
        ],
      },
    ],
    "recipient": 1000,
  },
]
`;

exports[`Shiden & AssetHub > 002: Shiden transfer USDT to AssetHub > 002: statemine event 1`] = `
[
  {
    "data": {
      "id": "(hash)",
      "origin": {
        "Sibling": "(rounded 2000)",
      },
      "success": true,
      "weightUsed": {
        "proofSize": "(rounded 7200)",
        "refTime": "(rounded 330000000)",
      },
    },
    "method": "Processed",
    "section": "messageQueue",
  },
]
`;
